File: Java/src/BlockPlacer.java
package net.glitchtechs.starapi;

import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scheduler.BukkitTask;

import java.util.concurrent.atomic.AtomicInteger;

public class BlockPlacer {
    private final JavaPlugin plugin;
    private Material material;
    private Direction direction;
    private int length;
    private int height;
    private long delayBetweenBlocks;
    private BukkitTask task;

    public BlockPlacer(JavaPlugin plugin) {
        this.plugin = plugin;
    }

    public static BlockPlacer create(JavaPlugin plugin, Material material) {
        BlockPlacer blockPlacer = new BlockPlacer(plugin);
        blockPlacer.material = material;
        return blockPlacer;
    }

    public BlockPlacer setDirection(Direction direction) {
        this.direction = direction;
        return this;
    }

    public BlockPlacer setLength(int length) {
        this.length = length;
        return this;
    }

    public BlockPlacer setHeight(int height) {
        this.height = height;
        return this;
    }

    public BlockPlacer setDelayBetweenBlocks(long delay) {
        this.delayBetweenBlocks = delay;
        return this;
    }

    public void build(Location origin) {
        World world = origin.getWorld();
        AtomicInteger x = new AtomicInteger(origin.getBlockX());
        AtomicInteger y = new AtomicInteger(origin.getBlockY());
        AtomicInteger z = new AtomicInteger(origin.getBlockZ());

        task = Bukkit.getScheduler().runTaskTimer(plugin, () -> {
            for (int j = 0; j < height; j++) {
                Block block = world.getBlockAt(x.get(), y.get() + j, z.get());
                block.setType(material);
            }

            switch (direction) {
                case EAST:
                    x.getAndIncrement();
                    break;
                case WEST:
                    x.getAndDecrement();
                    break;
                case SOUTH:
                    z.getAndIncrement();
                    break;
                case NORTH:
                    z.getAndDecrement();
                    break;
                case UP:
                    y.getAndIncrement();
                    break;
                case DOWN:
                    y.getAndDecrement();
                    break;
                case NORTHEAST:
                    x.getAndIncrement();
                    z.getAndDecrement();
                    break;
                case SOUTHEAST:
                    x.getAndIncrement();
                    z.getAndIncrement();
                    break;
                case SOUTHWEST:
                    x.getAndDecrement();
                    z.getAndIncrement();
                    break;
                case NORTHWEST:
                    x.getAndDecrement();
                    z.getAndDecrement();
                    break;
            }

            length--;

            if (length <= 0) {
                task.cancel();
            }
        }, delayBetweenBlocks, delayBetweenBlocks);
    }

    public enum Direction {
        EAST, WEST, SOUTH, NORTH, UP, DOWN, NORTHEAST, SOUTHEAST, SOUTHWEST, NORTHWEST
    }
}

File: Kotlin/src/BlockPlacer.kt
package net.glitchtechs.starapi

import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.scheduler.BukkitTask
import java.util.concurrent.atomic.AtomicInteger


class BlockPlacer(private val plugin: JavaPlugin) {
    private var material: Material? = null
    private var direction: Direction? = null
    private var length = 0
    private var height = 0
    private var delayBetweenBlocks: Long = 0
    private var task: BukkitTask? = null

    fun setDirection(direction: Direction?): BlockPlacer {
        this.direction = direction
        return this
    }

    fun setLength(length: Int): BlockPlacer {
        this.length = length
        return this
    }

    fun setHeight(height: Int): BlockPlacer {
        this.height = height
        return this
    }

    fun setDelayBetweenBlocks(delay: Long): BlockPlacer {
        this.delayBetweenBlocks = delay
        return this
    }

    fun build(origin: Location) {
        val world = origin.world
        val x = AtomicInteger(origin.blockX)
        val y = AtomicInteger(origin.blockY)
        val z = AtomicInteger(origin.blockZ)

        task = Bukkit.getScheduler().runTaskTimer(plugin, Runnable {
            for (j in 0 until height) {
                val block = world!!.getBlockAt(x.get(), y.get() + j, z.get())
                block.type = material!!
            }
            when (direction) {
                Direction.EAST -> x.getAndIncrement()
                Direction.WEST -> x.getAndDecrement()
                Direction.SOUTH -> z.getAndIncrement()
                Direction.NORTH -> z.getAndDecrement()
                Direction.UP -> y.getAndIncrement()
                Direction.DOWN -> y.getAndDecrement()
                Direction.NORTHEAST -> {
                    x.getAndIncrement()
                    z.getAndDecrement()
                }

                Direction.SOUTHEAST -> {
                    x.getAndIncrement()
                    z.getAndIncrement()
                }

                Direction.SOUTHWEST -> {
                    x.getAndDecrement()
                    z.getAndIncrement()
                }

                Direction.NORTHWEST -> {
                    x.getAndDecrement()
                    z.getAndDecrement()
                }

                null -> TODO()
            }
            length--
            if (length <= 0) {
                task!!.cancel()
            }
        }, delayBetweenBlocks, delayBetweenBlocks)
    }

    enum class Direction {
        EAST, WEST, SOUTH, NORTH, UP, DOWN, NORTHEAST, SOUTHEAST, SOUTHWEST, NORTHWEST
    }

    companion object {
        fun create(plugin: JavaPlugin, material: Material?): BlockPlacer {
            val blockPlacer = BlockPlacer(plugin)
            blockPlacer.material = material
            return blockPlacer
        }
    }
}
